Sources/Serializer/CodableSerializer.swift

```diff
@@ -0,0 +1,69 @@
+//
+//  CodableSerializer.swift
+//  Cara
+//
+//  Created by Jelle Vandebeeck on 02/12/2018.
+//
+
+import Foundation
+
+/// You can use this serializer when you want to convert a data object returned from the server and parse
+/// it to a `Codable` model.
+///
+/// How to use this:
+/// ```swift
+/// let serializer = CodableSerializer<User>()
+/// ```
+/// In the example above the `User` conforms to `Codable`.
+public struct CodableSerializer<Model: Codable>: Serializer {
+
+    // MARK: - Internal
+
+    private var decoder: JSONDecoder
+
+    // MARK: - Init
+
+    /// You can customise the `CodableSerializer` by passing another `JSONDecoder` is needed.
+    ///
+    /// - parameter decoder: The decoder what will be used to convert the json.
+    init(decoder: JSONDecoder = JSONDecoder()) {
+        self.decoder = decoder
+    }
+    
+    // MARK: - Response
+    
+    /// The codable response returned by the `CodableSerializer`.
+    public enum Response {
+        /// When data is correctly parsed we return the `Codable` instance. This can also be nil when no
+        /// data was returned.
+        case success(Model?)
+        /// Failure is triggered an error is returned or when parsing the data didn't succeed.
+        case failure(Error)
+    }
+    
+    // MARK: - Serialize
+    
+    public func serialize(data: Data?, error: Error?, response: URLResponse?) -> CodableSerializer<Model>.Response {
+        // When an error occurs we return this error.
+        if let error = error { return .failure(error) }
+        
+        // When a status code larger or equal than 400 is returned we return a custom error.
+        if
+            let response = response,
+            response.httpStatusCode >= 400 {
+            return .failure(ResponseError.httpError(statusCode: response.httpStatusCode))
+        }
+        
+        /// When no data object (or an empty one) is returned from the server we have a successful.
+        guard
+            let data = data,
+            !data.isEmpty else { return .success(nil) }
```

Comment:
Kunnen we hier expliciet niet de response code teruggeven? Bv als je een put doet krijg je een 201 (dacht ik) maar zo weet je dat verder in de app ook als een put bv succesvol is of wil je daar niet vanuit gaan?

Reply:
Ik vind dat niet echt nodig voor de standaard implementatie, stel dat je iets custom wil afhandelen, dan dat nog altijd door een andere `Serializer` te maken. Er zijn zo te veel gevallen denk ik, maar je de mogelijkheid is er. Ik wou vooral een basis implementatie hebben.